File: cloud-admin-portal/backend/app/services/pg_client.py
"""pg-tikv PostgreSQL client for tenant/user/observability operations.

Use a pure-Python driver (`pg8000`) to avoid spawning `psql` subprocesses.
"""

import time
from contextlib import contextmanager
from dataclasses import dataclass, field
from typing import List, Optional, Sequence, Tuple

import pg8000.dbapi as pg_dbapi
import sqlparse

from ..models import ObservabilitySummary, QuerySample


def _quote_ident(ident: str) -> str:
    # Minimal, safe SQL identifier quoting.
    # pg-tikv uses sqlparser, so standard Postgres quoting is expected to work.
    return '"' + ident.replace('"', '""') + '"'


def _format_cell(value: object) -> str:
    if value is None:
        return ""
    if isinstance(value, bool):
        return "t" if value else "f"
    return str(value)


def _format_row_pipe(row: Sequence[object]) -> str:
    return "|".join(_format_cell(v) for v in row)


@dataclass
class UserInfo:
    """User information from pg_roles."""
    
    name: str
    is_superuser: bool
    can_login: bool
    can_create_db: bool
    can_create_role: bool
    roles: List[str] = field(default_factory=list)

class PgTikvClient:
    """Client for pg-tikv operations (users + observability + ad-hoc SQL).

    Uses `pg8000` (pure Python).
    """
    
    def __init__(self, host: str, port: int):
        """Initialize pg-tikv client.
        
        Args:
            host: pg-tikv server host
            port: pg-tikv server port
        """
        self.host = host
        self.port = port
    
    @contextmanager
    def _connect(self, tenant: str, user: str, password: str) -> object:
        connect_user = f"{tenant}.{user}"
        conn = pg_dbapi.connect(
            user=connect_user,
            password=password,
            host=self.host,
            port=self.port,
            database="postgres",
        )
        try:
            conn.autocommit = True  # type: ignore[attr-defined]
        except Exception:
            pass

        try:
            yield conn
        finally:
            try:
                conn.close()  # type: ignore[attr-defined]
            except Exception:
                pass

    def _run_sql(
        self,
        tenant: str,
        user: str,
        password: str,
        sql: str
    ) -> Tuple[str, str, int]:
        """Execute SQL and return a `psql -t -A`-compatible output string.
        
        Supports multi-statement SQL scripts by splitting on semicolons
        using sqlparse for proper handling of strings and comments.
        
        Args:
            tenant: Tenant name
            user: Username
            password: Password
            sql: SQL to execute (can contain multiple statements)
        
        Returns:
            Tuple of (stdout, stderr, return_code)
        """
        statements = sqlparse.split(sql)
        statements = [s.strip() for s in statements if s.strip()]
        
        if not statements:
            return "", "", 0
        
        out_lines: List[str] = []
        try:
            with self._connect(tenant, user, password) as conn:
                cursor = conn.cursor()  # type: ignore[attr-defined]
                
                for stmt in statements:
                    if not stmt or stmt.startswith("--"):
                        continue
                    
                    cursor.execute(stmt)  # type: ignore[attr-defined]
                    
                    if getattr(cursor, "description", None) is not None:
                        rows = cursor.fetchall()  # type: ignore[attr-defined]
                        for row in rows:
                            out_lines.append(_format_row_pipe(row))
                        
                        if out_lines and statements.index(stmt) < len(statements) - 1:
                            out_lines.append("")
                
                try:
                    cursor.close()  # type: ignore[attr-defined]
                except Exception:
                    pass

            return "\n".join(out_lines).strip(), "", 0
        except Exception as e:
            return "", str(e), 1
    
    def execute_sql(
        self,
        tenant: str,
        user: str,
        password: str,
        sql: str
    ) -> Optional[str]:
        """Execute SQL and return result.
        
        Args:
            tenant: Tenant name
            user: Username
            password: Password
            sql: SQL to execute
        
        Returns:
            Result string or None on error
        """
        stdout, stderr, rc = self._run_sql(tenant, user, password, sql)
        if rc != 0:
            return None
        return stdout
    
    def test_connection(self, tenant: str, user: str, password: str) -> bool:
        """Test if credentials are valid.
        
        Args:
            tenant: Tenant name
            user: Username
            password: Password
        
        Returns:
            True if connection succeeds
        """
        result = self.execute_sql(tenant, user, password, "SELECT 1")
        return result == "1"

    def get_observability_summary(
        self,
        tenant: str,
        user: str,
        password: str,
    ) -> Tuple[Optional[ObservabilitySummary], Optional[str]]:
        sql = """
            SELECT
                window_seconds,
                statement_count,
                txn_commit_count,
                error_count,
                qps,
                tps,
                latency_avg_ms,
                latency_p99_ms,
                active_connections
            FROM _pgtikv_sys_observability()
        """
        stdout, stderr, rc = self._run_sql(tenant, user, password, sql)
        if rc != 0:
            return None, stderr or "Query execution failed"
        if not stdout:
            return None, "Empty response from pg-tikv"

        parts = stdout.split("|")
        if len(parts) < 9:
            return None, f"Unexpected response: {stdout!r}"

        try:
            summary = ObservabilitySummary(
                window_seconds=int(parts[0]),
                statement_count=int(parts[1]),
                txn_commit_count=int(parts[2]),
                error_count=int(parts[3]),
                qps=float(parts[4]),
                tps=float(parts[5]),
                latency_avg_ms=float(parts[6]),
                latency_p99_ms=float(parts[7]),
                active_connections=int(parts[8]),
            )
        except ValueError as e:
            return None, f"Failed to parse pg-tikv response: {e}"

        return summary, None

    def get_observability_samples(
        self,
        tenant: str,
        user: str,
        password: str,
    ) -> Tuple[List[QuerySample], Optional[str]]:
        sql = """
            SELECT
                query,
                sample_count,
                error_count,
                latency_avg_ms,
                latency_p99_ms,
                latency_max_ms,
                last_seen_ms_ago
            FROM _pgtikv_sys_query_samples()
        """
        stdout, stderr, rc = self._run_sql(tenant, user, password, sql)
        if rc != 0:
            return [], stderr or "Query execution failed"
        if not stdout:
            return [], None

        samples: List[QuerySample] = []
        for line in stdout.splitlines():
            if not line.strip():
                continue
            parts = line.split("|")
            if len(parts) < 7:
                continue
            try:
                samples.append(
                    QuerySample(
                        query=parts[0],
                        sample_count=int(parts[1]),
                        error_count=int(parts[2]),
                        latency_avg_ms=float(parts[3]),
                        latency_p99_ms=float(parts[4]),
                        latency_max_ms=float(parts[5]),
                        last_seen_ms_ago=int(parts[6]),
                    )
                )
            except ValueError:
                continue

        return samples, None
    
    def list_users(
        self,
        tenant: str,
        admin_user: str,
        admin_password: str
    ) -> List[UserInfo]:
        """List all users in the tenant.
        
        Args:
            tenant: Tenant name
            admin_user: Admin username
            admin_password: Admin password
        
        Returns:
            List of UserInfo objects
        """
        sql = """
            SELECT rolname, rolsuper, rolcanlogin, rolcreatedb, rolcreaterole
            FROM pg_catalog.pg_roles
            WHERE rolname NOT LIKE 'pg_%'
        """
        result = self.execute_sql(tenant, admin_user, admin_password, sql)
        
        if not result:
            # Return at least the admin user if query fails
            return [UserInfo(
                name=admin_user,
                is_superuser=True,
                can_login=True,
                can_create_db=True,
                can_create_role=True,
            )]
        
        users = []
        for line in result.split("\n"):
            if not line.strip():
                continue
            parts = line.split("|")
            if len(parts) >= 5:
                users.append(UserInfo(
                    name=parts[0],
                    is_superuser=parts[1].lower() == "t",
                    can_login=parts[2].lower() == "t",
                    can_create_db=parts[3].lower() == "t",
                    can_create_role=parts[4].lower() == "t",
                ))
        
        return users if users else [UserInfo(
            name=admin_user,
            is_superuser=True,
            can_login=True,
            can_create_db=True,
            can_create_role=True,
        )]
    
    def create_user(
        self,
        tenant: str,
        admin_user: str,
        admin_password: str,
        new_user: str,
        new_password: str,
        superuser: bool = False
    ) -> bool:
        """Create a new user in the tenant.
        
        Args:
            tenant: Tenant name
            admin_user: Admin username
            admin_password: Admin password
            new_user: New username
            new_password: New user's password
            superuser: Grant superuser privileges
        
        Returns:
            True if created successfully
        """
        options = "SUPERUSER" if superuser else ""
        # Escape single quotes in password
        escaped_password = new_password.replace("'", "''")
        sql = (
            f"CREATE ROLE {_quote_ident(new_user)} WITH LOGIN PASSWORD '{escaped_password}' {options}"
        )
        result = self.execute_sql(tenant, admin_user, admin_password, sql)
        return result is not None
    
    def drop_user(
        self,
        tenant: str,
        admin_user: str,
        admin_password: str,
        username: str
    ) -> bool:
        """Delete a user from the tenant.
        
        Args:
            tenant: Tenant name
            admin_user: Admin username
            admin_password: Admin password
            username: User to delete
        
        Returns:
            True if deleted successfully
        """
        sql = f"DROP ROLE IF EXISTS {_quote_ident(username)}"
        result = self.execute_sql(tenant, admin_user, admin_password, sql)
        return result is not None
    
    def reset_password(
        self,
        tenant: str,
        admin_user: str,
        admin_password: str,
        target_user: str,
        new_password: str
    ) -> bool:
        """Reset a user's password.
        
        Args:
            tenant: Tenant name
            admin_user: Admin username
            admin_password: Admin password
            target_user: User whose password to reset
            new_password: New password
        
        Returns:
            True if reset successfully
        """
        # Escape single quotes in password
        escaped_password = new_password.replace("'", "''")
        sql = f"ALTER ROLE {_quote_ident(target_user)} WITH PASSWORD '{escaped_password}'"
        result = self.execute_sql(tenant, admin_user, admin_password, sql)
        return result is not None

    def bootstrap_admin_password(
        self,
        tenant: str,
        admin_user: str,
        new_password: str,
        default_password: str = "admin",
        max_retries: int = 5
    ) -> bool:
        """Set admin password for a newly created tenant.
        
        When a new keyspace is created, pg-tikv bootstraps an admin user
        with default password "admin" on first connection. This method:
        1. Connects with default password to trigger bootstrap
        2. Changes password to the desired value
        
        Args:
            tenant: Tenant name (keyspace)
            admin_user: Admin username (typically "admin")
            new_password: Desired admin password
            default_password: Default bootstrap password (default: "admin")
            max_retries: Number of connection attempts (keyspace creation may take time)
        
        Returns:
            True if password was set successfully
        """
        for attempt in range(max_retries):
            if self.test_connection(tenant, admin_user, default_password):
                break
            time.sleep(0.5)
        else:
            return False
        
        escaped_password = new_password.replace("'", "''")
        sql = f"ALTER ROLE {admin_user} WITH PASSWORD '{escaped_password}'"
        result = self.execute_sql(tenant, admin_user, default_password, sql)
        return result is not None

File: cloud-admin-portal/backend/src/services/pg_client.rs
use crate::models::{ColumnInfo, SqlResult, UserResponse};
use serde_json::Value;
use tokio_postgres::types::Type;
use tokio_postgres::NoTls;

pub struct PgClient {
    host: String,
    port: u16,
}

/// Validate a SQL identifier (username/role name).
/// Only allows alphanumeric + underscore, must be non-empty, max 63 chars.
fn validate_identifier(name: &str) -> Result<(), String> {
    if name.is_empty() {
        return Err("Identifier cannot be empty".into());
    }
    if name.len() > 63 {
        return Err("Identifier too long (max 63 chars)".into());
    }
    if !name.chars().all(|c| c.is_ascii_alphanumeric() || c == '_') {
        return Err(format!(
            "Invalid identifier '{}': only [a-zA-Z0-9_] allowed",
            name
        ));
    }
    Ok(())
}

/// Escape a value for use inside a SQL single-quoted string literal.
/// Doubles any single-quote characters.
fn escape_sql_string(s: &str) -> String {
    s.replace('\'', "''")
}

/// Escape a value for use in a libpq connection string.
/// Wraps in single quotes, escaping backslashes and single quotes.
fn escape_connstr_value(s: &str) -> String {
    format!("'{}'", s.replace('\\', "\\\\").replace('\'', "\\'"))
}

fn split_sql_statements(sql: &str) -> Vec<String> {
    let bytes = sql.as_bytes();
    let mut statements = Vec::new();
    let mut in_single = false;
    let mut in_double = false;
    let mut in_line_comment = false;
    let mut in_block_comment = false;
    let mut dollar_delim: Option<Vec<u8>> = None;
    let mut start = 0usize;
    let mut i = 0usize;

    while i < bytes.len() {
        if let Some(ref delim) = dollar_delim {
            let dlen = delim.len();
            if i + dlen <= bytes.len() && bytes[i..i + dlen] == *delim.as_slice() {
                dollar_delim = None;
                i += dlen;
            } else {
                i += 1;
            }
            continue;
        }

        if in_line_comment {
            if bytes[i] == b'\n' {
                in_line_comment = false;
            }
            i += 1;
            continue;
        }

        if in_block_comment {
            if bytes[i] == b'*' && i + 1 < bytes.len() && bytes[i + 1] == b'/' {
                in_block_comment = false;
                i += 2;
            } else {
                i += 1;
            }
            continue;
        }

        let b = bytes[i];

        if !in_single && !in_double {
            if b == b'-' && i + 1 < bytes.len() && bytes[i + 1] == b'-' {
                in_line_comment = true;
                i += 2;
                continue;
            }
            if b == b'/' && i + 1 < bytes.len() && bytes[i + 1] == b'*' {
                in_block_comment = true;
                i += 2;
                continue;
            }
        }

        if b == b'\'' && !in_double {
            in_single = !in_single;
            i += 1;
            continue;
        }

        if b == b'"' && !in_single {
            in_double = !in_double;
            i += 1;
            continue;
        }

        if !in_single && !in_double {
            if b == b'$' {
                let prev_ok =
                    i == 0 || !(bytes[i - 1].is_ascii_alphanumeric() || bytes[i - 1] == b'_');
                if prev_ok {
                    if i + 1 < bytes.len() && bytes[i + 1] == b'$' {
                        dollar_delim = Some(b"$$".to_vec());
                        i += 2;
                        continue;
                    }
                    let mut j = i + 1;
                    while j < bytes.len() && (bytes[j].is_ascii_alphanumeric() || bytes[j] == b'_')
                    {
                        j += 1;
                    }
                    if j > i + 1 && j < bytes.len() && bytes[j] == b'$' {
                        dollar_delim = Some(bytes[i..=j].to_vec());
                        i = j + 1;
                        continue;
                    }
                }
            }

            if b == b';' {
                let stmt = sql[start..i].trim();
                if !stmt.is_empty() {
                    statements.push(stmt.to_string());
                }
                start = i + 1;
                i += 1;
                continue;
            }
        }

        i += 1;
    }

    let stmt = sql[start..].trim();
    if !stmt.is_empty() {
        statements.push(stmt.to_string());
    }

    statements
}

fn detect_command(sql: &str) -> String {
    let compact = sql.split_whitespace().collect::<Vec<_>>();
    if compact.is_empty() {
        return "UNKNOWN".to_string();
    }
    let first = compact[0].to_ascii_uppercase();
    if matches!(first.as_str(), "CREATE" | "ALTER" | "DROP") && compact.len() >= 2 {
        format!("{} {}", first, compact[1].to_ascii_uppercase())
    } else {
        first
    }
}

fn column_type_name(column_type: &Type) -> String {
    match *column_type {
        Type::BOOL => "boolean".to_string(),
        Type::INT2 => "smallint".to_string(),
        Type::INT4 => "integer".to_string(),
        Type::INT8 => "bigint".to_string(),
        Type::FLOAT4 => "real".to_string(),
        Type::FLOAT8 => "double precision".to_string(),
        Type::NUMERIC => "numeric".to_string(),
        Type::TEXT => "text".to_string(),
        Type::VARCHAR => "character varying".to_string(),
        Type::BPCHAR => "character".to_string(),
        Type::BYTEA => "bytea".to_string(),
        Type::DATE => "date".to_string(),
        Type::TIMESTAMP => "timestamp".to_string(),
        Type::TIMESTAMPTZ => "timestamp with time zone".to_string(),
        Type::TIME => "time".to_string(),
        Type::TIMETZ => "time with time zone".to_string(),
        Type::UUID => "uuid".to_string(),
        Type::JSON => "json".to_string(),
        Type::JSONB => "jsonb".to_string(),
        Type::UNKNOWN => "unknown".to_string(),
        _ => column_type.name().to_string(),
    }
}

fn text_to_json(val: Option<&str>, ty: &Type) -> Value {
    let s = match val {
        Some(v) => v,
        None => return Value::Null,
    };
    match *ty {
        Type::BOOL => Value::from(s.eq_ignore_ascii_case("t") || s.eq_ignore_ascii_case("true")),
        Type::INT2 | Type::INT4 => s
            .parse::<i64>()
            .map_or(Value::from(s.to_string()), Value::from),
        Type::INT8 => s
            .parse::<i64>()
            .map_or(Value::from(s.to_string()), Value::from),
        Type::FLOAT4 | Type::FLOAT8 | Type::NUMERIC => s
            .parse::<f64>()
            .map_or(Value::from(s.to_string()), Value::from),
        _ => Value::from(s.to_string()),
    }
}

fn format_pg_error(e: &tokio_postgres::Error) -> String {
    if let Some(db_err) = e.as_db_error() {
        let mut parts = vec![format!("{}: {}", db_err.severity(), db_err.message())];
        if let Some(detail) = db_err.detail() {
            parts.push(format!("DETAIL: {detail}"));
        }
        if let Some(hint) = db_err.hint() {
            parts.push(format!("HINT: {hint}"));
        }
        let code = db_err.code().code();
        if !code.is_empty() {
            parts.push(format!("SQLSTATE: {code}"));
        }
        parts.join("\n")
    } else {
        e.to_string()
    }
}

fn query_like_statement(sql: &str) -> bool {
    let upper = sql.trim_start().to_ascii_uppercase();
    upper.starts_with("SELECT")
        || upper.starts_with("WITH")
        || upper.starts_with("SHOW")
        || upper.starts_with("VALUES")
        || upper.contains(" RETURNING ")
}

impl PgClient {
    pub fn new(host: &str, port: u16) -> Self {
        Self {
            host: host.to_string(),
            port,
        }
    }

    async fn connect(
        &self,
        tenant_id: &str,
        user: &str,
        password: &str,
    ) -> Result<tokio_postgres::Client, String> {
        let connstr = format!(
            "host={} port={} user={} password={} dbname=postgres",
            escape_connstr_value(&self.host),
            self.port,
            escape_connstr_value(&format!("{}.{}", tenant_id, user)),
            escape_connstr_value(password),
        );
        let (client, conn) = tokio_postgres::connect(&connstr, NoTls)
            .await
            .map_err(|e| e.to_string())?;
        tokio::spawn(async move {
            conn.await.ok();
        });
        Ok(client)
    }

    pub async fn test_connection(&self, keyspace: &str, user: &str, password: &str) -> bool {
        match self.connect(keyspace, user, password).await {
            Ok(c) => c.simple_query("SELECT 1").await.is_ok(),
            Err(_) => false,
        }
    }

    pub async fn bootstrap_admin_password(
        &self,
        keyspace: &str,
        user: &str,
        default_password: &str,
        desired_password: &str,
    ) -> bool {
        // If desired password already works, nothing to do
        if self.test_connection(keyspace, user, desired_password).await {
            return true;
        }

        // Wait for keyspace to become available, then connect with default password
        for attempt in 0..5 {
            if attempt > 0 {
                tokio::time::sleep(std::time::Duration::from_millis(500)).await;
            }
            if self.test_connection(keyspace, user, default_password).await {
                // Change password to desired one
                return self
                    .reset_password(keyspace, user, default_password, user, desired_password)
                    .await;
            }
        }
        false
    }

    pub async fn bootstrap_default_extensions(&self, keyspace: &str, user: &str, password: &str) {
        const DEFAULT_EXTENSIONS: &[&str] = &["http", "fs9", "pg_cron"];
        let client = match self.connect(keyspace, user, password).await {
            Ok(c) => c,
            Err(e) => {
                tracing::warn!("bootstrap_default_extensions: connect failed: {e}");
                return;
            }
        };
        for ext in DEFAULT_EXTENSIONS {
            let sql = format!("CREATE EXTENSION IF NOT EXISTS \"{ext}\"");
            if let Err(e) = client.simple_query(&sql).await {
                tracing::warn!("bootstrap_default_extensions: {ext}: {e}");
            }
        }
    }

    pub async fn list_users(
        &self,
        tenant_id: &str,
        admin_user: &str,
        admin_password: &str,
    ) -> Vec<UserResponse> {
        let client = match self.connect(tenant_id, admin_user, admin_password).await {
            Ok(c) => c,
            Err(e) => {
                tracing::warn!("list_users connect failed: {e}");
                return Vec::new();
            }
        };
        match client
            .query(
                "SELECT rolname, rolsuper, rolcanlogin, rolcreatedb, rolcreaterole FROM pg_roles",
                &[],
            )
            .await
        {
            Ok(rows) => rows
                .iter()
                .map(|r| UserResponse {
                    name: r.get::<_, String>(0),
                    is_superuser: r.get::<_, bool>(1),
                    can_login: r.get::<_, bool>(2),
                    can_create_db: r.get::<_, bool>(3),
                    can_create_role: r.get::<_, bool>(4),
                })
                .collect(),
            Err(e) => {
                tracing::warn!("list_users query failed: {e}");
                Vec::new()
            }
        }
    }

    pub async fn create_user(
        &self,
        keyspace: &str,
        admin_user: &str,
        admin_password: &str,
        new_user: &str,
        new_password: &str,
        superuser: bool,
    ) -> bool {
        if let Err(e) = validate_identifier(new_user) {
            tracing::warn!("create_user rejected: {e}");
            return false;
        }
        let client = match self.connect(keyspace, admin_user, admin_password).await {
            Ok(c) => c,
            Err(_) => return false,
        };
        let su = if superuser { " SUPERUSER" } else { "" };
        let sql = format!(
            "CREATE ROLE {new_user} WITH LOGIN PASSWORD '{}'{su}",
            escape_sql_string(new_password),
        );
        client.simple_query(&sql).await.is_ok()
    }

    pub async fn drop_user(
        &self,
        keyspace: &str,
        admin_user: &str,
        admin_password: &str,
        username: &str,
    ) -> bool {
        if let Err(e) = validate_identifier(username) {
            tracing::warn!("drop_user rejected: {e}");
            return false;
        }
        let client = match self.connect(keyspace, admin_user, admin_password).await {
            Ok(c) => c,
            Err(_) => return false,
        };
        client
            .simple_query(&format!("DROP ROLE {username}"))
            .await
            .is_ok()
    }

    pub async fn reset_password(
        &self,
        keyspace: &str,
        admin_user: &str,
        admin_password: &str,
        target_user: &str,
        new_password: &str,
    ) -> bool {
        if let Err(e) = validate_identifier(target_user) {
            tracing::warn!("reset_password rejected: {e}");
            return false;
        }
        let client = match self.connect(keyspace, admin_user, admin_password).await {
            Ok(c) => c,
            Err(_) => return false,
        };
        let sql = format!(
            "ALTER ROLE {target_user} WITH PASSWORD '{}'",
            escape_sql_string(new_password),
        );
        client.simple_query(&sql).await.is_ok()
    }

    pub async fn run_sql(
        &self,
        keyspace: &str,
        user: &str,
        password: &str,
        sql: &str,
    ) -> Result<String, String> {
        let client = self.connect(keyspace, user, password).await?;
        let rows = client
            .simple_query(sql)
            .await
            .map_err(|e| format_pg_error(&e))?;
        let mut output = String::new();
        for msg in rows {
            if let tokio_postgres::SimpleQueryMessage::Row(row) = msg {
                let cols: Vec<String> = (0..row.len())
                    .map(|i| row.get(i).unwrap_or("").to_string())
                    .collect();
                output.push_str(&cols.join("|"));
                output.push('\n');
            }
        }
        Ok(output)
    }

    /// Execute SQL and return structured results.
    ///
    /// Uses prepare() for column type metadata and simple_query() for data
    /// retrieval (text format) to avoid binary encoding mismatches with pg-tikv.
    pub async fn run_sql_structured(
        &self,
        keyspace: &str,
        user: &str,
        password: &str,
        sql: &str,
    ) -> Result<SqlResult, String> {
        let client = self.connect(keyspace, user, password).await?;
        let statements = split_sql_statements(sql);
        if statements.is_empty() {
            return Err("Empty SQL query".to_string());
        }

        let mut last_result = SqlResult {
            columns: Vec::new(),
            rows: Vec::new(),
            row_count: 0,
            command: "UNKNOWN".to_string(),
            error: None,
        };

        for statement in &statements {
            let command = detect_command(statement);

            if query_like_statement(statement) {
                let col_types = match client.prepare(statement).await {
                    Ok(stmt) => stmt
                        .columns()
                        .iter()
                        .map(|c| (c.name().to_string(), c.type_().clone()))
                        .collect::<Vec<_>>(),
                    Err(e) => return Err(format_pg_error(&e)),
                };

                let messages = client
                    .simple_query(statement)
                    .await
                    .map_err(|e| format_pg_error(&e))?;

                let columns: Vec<ColumnInfo> = col_types
                    .iter()
                    .map(|(name, ty)| ColumnInfo {
                        name: name.clone(),
                        data_type: column_type_name(ty),
                    })
                    .collect();

                let mut out_rows: Vec<Vec<Value>> = Vec::new();
                for msg in messages {
                    if let tokio_postgres::SimpleQueryMessage::Row(row) = msg {
                        let vals: Vec<Value> = (0..col_types.len())
                            .map(|i| text_to_json(row.get(i), &col_types[i].1))
                            .collect();
                        out_rows.push(vals);
                    }
                }

                last_result = SqlResult {
                    row_count: out_rows.len(),
                    columns,
                    rows: out_rows,
                    command,
                    error: None,
                };
            } else {
                let messages = client
                    .simple_query(statement)
                    .await
                    .map_err(|e| format_pg_error(&e))?;

                let mut affected: u64 = 0;
                for msg in messages {
                    if let tokio_postgres::SimpleQueryMessage::CommandComplete(n) = msg {
                        affected = n;
                    }
                }

                last_result = SqlResult {
                    columns: Vec::new(),
                    rows: Vec::new(),
                    row_count: affected as usize,
                    command,
                    error: None,
                };
            }
        }

        Ok(last_result)
    }

    pub async fn get_observability_summary(
        &self,
        keyspace: &str,
        user: &str,
        password: &str,
    ) -> Result<Option<serde_json::Value>, String> {
        let out = self
            .run_sql(
                keyspace,
                user,
                password,
                "SELECT * FROM _pgtikv_sys_observability()",
            )
            .await?;
        if out.is_empty() {
            return Ok(None);
        }
        let parts: Vec<&str> = out.trim().split('|').collect();
        if parts.len() < 9 {
            return Ok(None);
        }
        Ok(Some(serde_json::json!({
            "window_seconds": parts[0].parse::<i64>().unwrap_or(0),
            "statement_count": parts[1].parse::<i64>().unwrap_or(0),
            "txn_commit_count": parts[2].parse::<i64>().unwrap_or(0),
            "error_count": parts[3].parse::<i64>().unwrap_or(0),
            "qps": parts[4].parse::<f64>().unwrap_or(0.0),
            "tps": parts[5].parse::<f64>().unwrap_or(0.0),
            "latency_avg_ms": parts[6].parse::<f64>().unwrap_or(0.0),
            "latency_p99_ms": parts[7].parse::<f64>().unwrap_or(0.0),
            "active_connections": parts[8].parse::<i64>().unwrap_or(0),
        })))
    }

    pub async fn get_observability_samples(
        &self,
        keyspace: &str,
        user: &str,
        password: &str,
    ) -> Vec<serde_json::Value> {
        match self
            .run_sql(
                keyspace,
                user,
                password,
                "SELECT * FROM _pgtikv_sys_query_samples()",
            )
            .await
        {
            Ok(out) => out
                .lines()
                .filter_map(|line| {
                    let parts: Vec<&str> = line.split('|').collect();
                    if parts.len() < 7 {
                        return None;
                    }
                    Some(serde_json::json!({
                        "query": parts[0],
                        "sample_count": parts[1].parse::<i64>().unwrap_or(0),
                        "error_count": parts[2].parse::<i64>().unwrap_or(0),
                        "latency_avg_ms": parts[3].parse::<f64>().unwrap_or(0.0),
                        "latency_p99_ms": parts[4].parse::<f64>().unwrap_or(0.0),
                        "latency_max_ms": parts[5].parse::<f64>().unwrap_or(0.0),
                        "last_seen_ms_ago": parts[6].parse::<i64>().unwrap_or(0),
                    }))
                })
                .collect(),
            Err(_) => Vec::new(),
        }
    }
}
